backends/ebpf/psa/ebpfPsaTable.cpp

```diff
@@ -564,4 +579,344 @@ bool EBPFTablePSA::dropOnNoMatchingEntryFound() const {
+        for (auto valueName : valueNames)
+            builder->appendFormat("&%s,", valueName);
+        builder->append("}");
+        builder->endOfStatement(true);
+
+        builder->newline();
+        builder->emitIndent();
+        builder->appendFormat("%s(%s, %s, &%s, &%s, &%s, &%s, %s, %s)",
+                              addPrefixFunctionName,
+                              cstring::to_cstring(samePrefixEntries.size()),
+                              cstring::to_cstring(tuple_id),
+                              tuplesMapName,
+                              prefixesMapName,
+                              keyMaskVarName,
+                              valueMask,
+                              keysArray,
+                              valuesArray);
+        builder->endOfStatement(true);
+
+        tuple_id++;
+    }
+}
+
+void EBPFTablePSA::emitKeysAndValues(CodeBuilder *builder,
+                                            std::vector<const IR::Entry *> &samePrefixEntries,
+                                            std::vector<cstring> &keyNames,
+                                            std::vector<cstring> &valueNames) {
+    CodeGenInspector cg(program->refMap, program->typeMap);
+    cg.setBuilder(builder);
+
+    for (auto entry : samePrefixEntries) {
+        cstring keyName = program->refMap->newName("key");
+        cstring valueName = program->refMap->newName("value");
+        keyNames.push_back(keyName);
+        valueNames.push_back(valueName);
+        // construct key
+        builder->emitIndent();
+        builder->appendFormat("struct %s %s = {}", keyTypeName.c_str(), keyName.c_str());
+        builder->endOfStatement(true);
+        for (size_t k = 0; k < keyGenerator->keyElements.size(); k++) {
+            auto keyElement = keyGenerator->keyElements[k];
+            cstring fieldName = get(keyFieldNames, keyElement);
+            CHECK_NULL(fieldName);
+            builder->emitIndent();
+            builder->appendFormat("%s.%s = ", keyName.c_str(), fieldName.c_str());
+            auto mtdecl = program->refMap->getDeclaration(keyElement->matchType->path, true);
+            auto matchType = mtdecl->getNode()->to<IR::Declaration_ID>();
+            auto expr = entry->keys->components[k];
+            auto ebpfType = get(keyTypes, keyElement);
+            if (auto km = expr->to<IR::Mask>()) {
+                km->left->apply(cg);
+                builder->append(" & ");
+                km->right->apply(cg);
+            } else {
+                expr->apply(cg);
+                builder->append(" & ");
+                unsigned width = 0;
+                if (auto hasWidth = ebpfType->to<IHasWidth>()) {
+                    width = hasWidth->widthInBits();
+                } else {
+                    BUG("Cannot assess field bit width");
+                }
+                builder->append("0x");
+                for (int j = 0; j < width / 8; j++)
```

Comment:
does this work with non-aligned widths too?

Reply:
you should factor this into a utility function getHexMask, or something of the kind.